client/executor/src/wasm_runtime.rs

```diff
@@ -343,6 +356,57 @@ fn decode_version(version: &[u8]) -> Result<RuntimeVersion, WasmError> {
 	}
 }
 
+fn decode_runtime_apis(mut apis: &[u8]) -> Result<Vec<([u8; 8], u32)>, WasmError> {
+	use std::convert::TryInto;
+
+	if apis.len() % 12 != 0 {
+		return Err(WasmError::Other(format!(
+			"the runtime apis section doesn't have a correct size"
+		)));
+	}
+
+	let mut result = Vec::with_capacity(apis.len() / 12);
+	while apis.len() != 0 {
+		let id: [u8; 8] = apis[0..8]
+			.try_into()
+			.expect("the source slice size is equal to the dest array length; qed");
+
+		let version = u32::from_le_bytes(
+			apis[8..12]
+				.try_into()
+				.expect("the source slice size is equal to the array length; qed"),
+		);
+
+		result.push((id, version));
+
+		apis = &apis[12..];
+	}
+
+	Ok(result)
+}
+
+/// Take the runtime blob and scan it for the custom wasm sections containing the version information
+/// and construct the `RuntimeVersion` from them.
+///
+/// If there are no such sections, it returns `None`. If there is an error during decoding those
+/// sections, `Err` will be returned.
+pub(crate) fn read_embedded_version(
+	blob: &RuntimeBlob,
+) -> Result<Option<RuntimeVersion>, WasmError> {
+	if let Some(version_section) = blob.custom_section_contents("runtime_version") {
+		let mut decoded_version = decode_version(&version_section)?;
```

Comment:
Here is your bug :D

It will first try to decode the old `RuntimeVersion` and after that, it will check the version of the `Core` runtime api. In our case here the runtime api doesn't exist and we will not try to decode the new `RuntimeVersion` with the correct transaction version.

Reply:
Ahh, good spot!